torch/_inductor/pattern_matcher.py

```diff
@@ -641,54 +726,60 @@ def pattern_eq(self, other: Any) -> bool:
 
 
 class MultiOutputPattern(PatternExpr):
-    def __init__(self, outputs):
+    outputs: List[Optional[PatternExpr]]
+
+    def __init__(self, outputs: Sequence[Optional[PatternExpr]]) -> None:
         super().__init__()
-        assert all(isinstance(x, (PatternExpr, type(None))) for x in outputs), outputs
-        self.outputs: List[Optional[PatternExpr]] = outputs
+        assert isinstance(outputs[0], _TargetExpr)
+        assert all(x is None or isinstance(x, PatternExpr) for x in outputs), outputs
+        self.outputs = list(outputs)
         self.op = outputs[0].op
 
     @property
-    def fns(self):
-        assert self.outputs[0] and hasattr(self.outputs[0], "fns")
-        return self.outputs[0].fns
+    def fns(self) -> Union[Callable[..., Any], str, Sequence[Any]]:
+        output = typing.cast(_TargetExpr, self.outputs[0])
```

Comment:
Could these protocol typing casts be settled with assert isinstance calls instead?

Reply:
We already verify output[0] on line 733 above during __init__ - so another assert here would be redundant. I'll add a comment to that effect.